design-proposals/cdrom-inject-eject.md

```diff
@@ -0,0 +1,173 @@
+# Overview
+Proposal for for adding inject/eject CD-ROM support to KubeVirt
+
+## Motivation
+In a world where all of our knowledge has been collected, indexed, and can be served up via a network connection, the humble CD-ROM is still a "thing." And to the dedicated hangers-on of this once ubiquitous technology, KubeVirt's CD-ROM support is lacking in a very fundamental way. Today, a user must reboot a VM in order to change the media in a CD-ROM. That is just. not. right. Fortunately, the virtualization technology that underpins KubeVirt has had this functionality for a long time and we can easily plug it into the KubeVirt API to unleash the full power of the CD-ROM
+
+## Goals
+- Extend existing APIs to support inject/eject CD-ROM operations
+- Extend virtctl to have cdrom specific commands
+- Support for PersistentVolumeClaim/DataVolume based disk images
+
+## Non Goals
+- Support for any volume type other than DavaVolume/PersistentVolumeClaim
+
+## Definition of Users
+KubeVirt end user, may not be a namespace/cluster admin
+
+## User Stories
+- As a VM owner, I want to be able to attach a CD-ROM for the purposes of installing an operating system, then be able eject the CD and perform a guest-side reboot to use the operating system.
+- As a VM Owner I want to attach an ISO/CD-ROM from my desktop to the VM without the need to reboot the VM
+- As a VM Owner I want to attach a PVC which contains an ISO to my VM without the need to reboot the VM
+- As a VM Owner I want to be able to "eject" (remove) the ISO/CD from my VM without the need to reboot the VM
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+Like most new features in KubeVirt, the building blocks already exist, just have to wire everything together
```

Comment:
Please consider alternative designs and their pros/cons. In particular, can you consider following https://github.com/kubevirt/community/blob/e4645eb/design-proposals/vm-rollout-strategy/vm-rollout-strategy.md to enable a simpler, declarative, API. According to it, the user expresses a VM spec that includes a CDROM with a specific PVC and the VM controller makes that happen.

Reply:
I can address that in the design, but just to answer you inline, I think the functionality you describe can be generalized and included in an all encompassing "Declarative Volume Hotplug" design

Reply:
> I can address that in the design,

Thanks.

> I think the functionality you describe can be generalized and included in an all encompassing "Declarative Volume Hotplug" design

Right, I'd love to see that. But I'd be satisfied in seeing CDROM-only as a first step.

Reply:
I think just CD-ROM being declarative would be confusing to users. 

To me, what is proposed here is a good first step in making this long awaited functionality available to users soon. Going full declarative is a lot more complicated. But is something I do agree that we should tackle eventually

Reply:
Please consider pros and cons each of these options in the text of the design.

Reply:
Adding my 2 cents here, declarative is a nice to have, however, this type of installation operation is imperative and kind of "manual". Do we really expect this to be used with git ops?